Enable specification of TLS Protocol Versions and Cipher Suites

jai1 opened the pull request:
### Motivation

Enable listening services (Broker, Proxy, Discovery Provider) to specify which TLS Protocol Versions and Cipher Suites they will negotiate with.
This is useful when you don't want the client to use downgraded protocols or ciphers which are weak.

### Modifications
Major code change is in SecurityUtility.createNettySslContextForServer rest is just config changes and tests.

### Result
Clients can't force servers (Broker, Proxy, Discovery Provider)  to use downgraded protocols or ciphers which are weak.

merlimat commented on pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:
```diff
@@ -179,7 +179,10 @@
     // Accept untrusted TLS certificate from client
     private boolean tlsAllowInsecureConnection = false;
-
+    // Specify the tls protocols and cipher the broker will use to negotiate during TLS Handshake.
```
add to `broker.conf`, with documentation and examples for both parameters

rdhabalia replied:
and in `proxy.conf` as well? should we have any test for this PR?

jai1 replied:
Added

maskit commented:
Huge +1. I’ve been worried about this.

IIRC, we pass a string literal “SSL” as an argument somewhere else, and probably we can pass “TLS1.2” instead. Maybe we should fix it too.

rdhabalia commented on pulsar-common/src/main/java/org/apache/pulsar/common/util/SecurityUtility.java:
```diff
@@ -95,12 +97,22 @@ public static SslContext createNettySslContextForClient(boolean allowInsecureCon
+        }
+
+        if (protocols != null && protocols.size() > 0) {
+            String[] protocolsArray = new String[protocols.size()];
```
`builder.protocols(protocols.toArray(protocolsArray));` ??

jai1 commented:
@rdhabalia - Have addressed your comments and added Tests
@merlimat - Have created a ticket for adding documentation - will do it tomorrow or so 
@maskit - Except for SecurityUtility.java  (line below) I didn't find any hardcoding - here too the code looks ok to me
```
SSLContext sslCtx = SSLContext.getInstance("TLS");
```
Can I get a +1 for this?

rdhabalia commented on pulsar-proxy/src/test/java/org/apache/pulsar/proxy/server/ProxyWithProxyAuthorizationTest.java:
```diff
@@ -306,6 +360,79 @@ public void textTlsHostVerificationProxyToBroker(boolean hostnameVerificationEna
+        proxyConfig.setTlsCiphers(tlsCiphers);
+        ProxyService proxyService = Mockito.spy(new ProxyService(proxyConfig));
+        proxyService.start();
+        Thread.sleep(1000);
```
umm..instead sleep can we check some condition until which we want to wait using `org.apache.pulsar.broker.auth.MockedPulsarServiceBaseTest.retryStrategically(..)`, it will help to avoid one more intermittent test failure. and we can also add test timeout=5Sec.

maskit commented:
@jai1 You may want to read these. If I understand correctly, HTTP entrypoints, which use `SecurityUtility::createSslContext`, may accept any protocol versions, even if we change the string to "TLS1.2".

https://docs.oracle.com/javase/9/docs/api/javax/net/ssl/SSLContext.html#getInstance-java.lang.String-
https://docs.oracle.com/javase/9/docs/specs/security/standard-names.html#sslcontext-algorithms
https://stackoverflow.com/questions/43481010/how-to-enable-only-tlsv1-2-on-java-8-server-application?noredirect=1&lq=1
